cmd/osbuild-image-tests/openstacktest/openstack.go

```diff
@@ -0,0 +1,136 @@
+// +build integration
+
+package openstacktest
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	"github.com/gophercloud/gophercloud/openstack/imageservice/v2/images"
+	"github.com/gophercloud/gophercloud/openstack/imageservice/v2/imagedata"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
+)
+
+func UploadImageToOpenStack(p *gophercloud.ProviderClient, imagePath string, imageName string) (*images.Image, error) {
+	client, err := openstack.NewImageServiceV2(p, gophercloud.EndpointOpts{
+		Region: os.Getenv("OS_REGION_NAME"),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Error creating ImageService client: %v", err)
+	}
+
+	// create a new image which gives us the ID
+	image, err := images.Create(client, images.CreateOpts{
+		Name: imageName,
+		DiskFormat: "qcow2",
+		ContainerFormat: "bare",
+	}).Extract()
+	if err != nil {
+		return image, fmt.Errorf("Creating image failed: %v", err)
+	}
+
+	// then upload the actual binary data
+	imageData, err := os.Open(imagePath)
+	if err != nil {
+		return image, fmt.Errorf("Error opening %s: %v", imagePath, err)
+	}
+	defer imageData.Close()
+
+	err = imagedata.Upload(client, image.ID, imageData).ExtractErr()
+	if err != nil {
+		return image, fmt.Errorf("Upload to OpenStack failed: %v", err)
+	}
+
+	// wait for the status to change from Queued to Active
+	err = gophercloud.WaitFor(300, func() (bool, error) {
```

Comment:
I'd bump this timeout much higher or remove it completely. Usually 5 min should be fine, but for AWS we had issues where this would suddenly take forever and our tests failing spuriously. Our test framework anyway has a global timeout that would cancel the test-run eventually.

Reply:
This function needs its timeout argument, I'll see what I can do. How much higher is higher enough ? 

FTR: if this test fails because the operation didn't complete in 5 minutes I am perfectly fine with restarting the test job anyway. Same goes for any other environment because they can fail for whatever reason outside our control.

Reply:
I'm happy to merge as is, but I'd make it 6 hours or so, and then rely on the global timeout instead. If it is genuinely stuck, then restarting isn't a problem, but most of the time it just takes a bit longer than we expected, and if we have enough of these we see a lot of spurious CI failures which would have been avoided by just waiting longer.